feat: customize infrastructure with plugins (#1418)
winglang/wing: docs/04-reference/winglang-plugins.md

This hook is useful for adding customizations that can not be applied in the
context of the preSynth hook. Its  worth noting that this is not meant as a
validation phase since the config is still mutable by subsequent plugins.

The following example manipulates the Terraform configuration to use a S3
backend. For brevity, the example uses  hard coded values. In practice these
could be configured through environment variables, class constructors,
configuration files, etc.
```js
// exports the postSynth function
exports.postSynth = function(config) {
  config.terraform.backend = {
    s3: {
      bucket: "my-wing-state-bucket",
      key: "plugins-rock.tfstate",
      region: "us-east-1",
    }  
  }
  return config;
}
```

### `validate` hook

API Reference
```ts
validate(config: any): void;
```

This hook is called right after the `postSynth` hook and provided the same
context object. In the context of a  Terraform-based target like `tf-aws`, this
is the same Terraform JSON configuration. However, does not allow configuration
to be mutated, which allows plugins to validate the configuration without
worrying about another plugin mutating after the fact.

> The following example validates that buckets all have versioning enabled if they
> don't it will throw an  error during compilation.

Reviewer: ```suggestion
The following example validates that buckets all have versioning enabled
and throw an error during compilation if they don't.
```

---
title: Compiler Plugins
id: compiler-plugins
description: Wing Compiler Plugins Reference
keywords: [Wing reference, Wing language, language, Wing language spec, Wing programming language, compiler plugins]
---

> Wing compiler plugins are used to extend the Wing compiler. They can be used to add non-functional customizations to the compilation's resulting infrastructure. For example, you can use a compiler plugin to enforce a tagging convention on all resources in the resulting infrastructure.

Reviewer 2: wrap to 100 width please ([rewrap] is amazing). Long lines make it harder to provide PR comments.

[rewrap]: https://marketplace.visualstudio.com/items?itemName=stkb.rewrap

Author: Oh my! Rewrap is amazing, my markdown life is forever changed 😁

> Wing compiler plugins are used to extend the Wing compiler. They can be used to add non-functional customizations to the compilation's resulting infrastructure. For example, you can use a compiler plugin to enforce a tagging convention on all resources in the resulting infrastructure.

Reviewer 2: ```suggestion
Wing compiler plugins are used to extend the Wing compiler. They can be used to customize the compilation output such as infrastructure definitions. For example, you can use a compiler plugin to enforce a tagging convention on all resources in the resulting infrastructure.
```

## Plugin Design

> A compiler plugin is a JavaScript file that exports multiple functions. These functions are called by the Wing compiler at various points during compilation also referred to as stages.

Reviewer 2: ```suggestion
A compiler plugin is a JavaScript file that exports multiple functions (called "hooks"). These hooks are called by the compiler at various points during compilation.
```

At each stage the compiler plugin is provided with a context object that allows the plugin to hook into the compilation process and apply customizations.

A single plugin can implement any number of stages. However, it is recommended to only implement the stages that are needed.

> ## Compilation Stages

Reviewer 2: I would use the term "hooks" instead of "stages". Hooks is more general purpose and doesn't necessarily imply sequential ordering. For example, we want to add a resource factory hook (#1365), and those are not "stages".

"The compiler offers hooks that can be used to customize its behavior at various points during the compilation."

Each compilation stage is meant to serve a specific purpose. The following is a list of the stages and their purpose:

> ### `preSynth` stage

Reviewer 2: ```suggestion
### `preSynth` hook
```

> 

Reviewer 2: For each hook, please add API reference (you can use TypeScript notation)

```js
preSynth(app: Construct): void;
```

> This stage is called before the Wing begins to synthesize. In the context of a cdktf target, this stage will have access to the root node of the construct tree. Allowing for the plugin to add or manipulate constructs before the synthesis process begins.

Reviewer: ```suggestion
This stage is called before the Wing begins to synthesize. In the context of a terraform-based target like `tf-aws`, this stage will have access to the root node of the construct tree. This allows the plugin to add or change [CDK for Terraform](https://github.com/hashicorp/terraform-cdk) constructs before the tree is synthesized into Terraform configuration.
```

> This stage is called before the Wing begins to synthesize. In the context of a terraform-based target like `tf-aws`, this stage will have access to the root node of the construct tree. This allows the plugin to add or change [CDK for Terraform](https://github.com/hashicorp/terraform-cdk) constructs before the tree is synthesized.

Reviewer 2: ```suggestion
This stage is called before the compiler begins to synthesize. In the context of a Terraform-based target like `tf-aws`, this stage will have access to the root node of the construct tree. This allows the plugin to add or change [CDK for Terraform](https://github.com/hashicorp/terraform-cdk) constructs before the tree is synthesized.
```

> The following example adds bucket to root node, note this plugin is intended to be used with the `tf-aws` target.

Reviewer: ```suggestion
The following example adds a bucket to the root node (note that this plugin is intended to be used with the `tf-aws` target).
```

> The following example adds a bucket to the root node (note this plugin is intended to be used with the `tf-aws` target).

Reviewer 2: > note this plugin is intended to be used with the `tf-aws` target.

Feels like we should offer some way for a plugin to validate that it is used within the right context. Let's open an issue to track this need?

Author: Yea I have an idea on how we can do this, Ill add it as a suggestion to the issue.

Author: https://github.com/winglang/wing/issues/1474

```js
const s3_bucket = require("@cdktf/provider-aws/lib/s3-bucket");

// exports the preSynth function
exports.preSynth = function(app) {
  // app is the root node of the construct tree
  new s3_bucket.S3Bucket(app, "MyPluginBucket", {
    bucket: "my-plugin-bucket",
  });
}
```

> ### `postSynth` stage
> 
> This stage is called after the Wing has finished synthesizing. In the context of a cdktf target, this stage will have access to the raw terraform JSON configuration. Allowing for manipulation of the JSON that is written to the compiled output directory.

Reviewer: On naming, I wonder if it makes sense to say that the `postSynth` **stage** is when all `postSynth` **hooks** (or **compiler hooks**) are called by the compiler. IMO it's not the stage that gets called, but the hook. Might be super pedantic though 😅

```suggestion
### `postSynth` stage

This stage runs after Wing has finished synthesizing artifacts. When compiling to a terraform-based target like `tf-aws`, the hook will have access to the raw terraform JSON configuration, allowing for manipulation of the JSON that is written to the compiled output directory.
```

Author: me word no good 😄 I agree it makes sense it has been updated

> This stage runs after the Wing has finished synthesizing artifacts. When compiling to a terraform-based target like `tf-aws`, the hook will have access to the raw terraform JSON configuration, allowing for manipulation of the JSON that is written to the compiled output directory.

Reviewer 2: ```suggestion
This stage runs after artifacts where synthesized. When compiling to a Terraform-based target like `tf-aws`, the hook will have access to the raw Terraform JSON configuration, allowing for manipulation of the JSON that is written to the compiled output directory.
```

This stage is useful for adding customizations that can not be applied in the context of the preSynth phase. Its worth noting that this is not meant as a validation phase since this stage is mutable by subsequent plugins.

> The following example manipulates the terraform configuration to use a s3 backend. For brevity, the example uses hard coded values. In practice these would be configurable.

Reviewer: ```suggestion
The following example manipulates the terraform configuration to use a S3 backend. For brevity, the example uses hard coded values. In practice these would be configurable.
```

Reviewer: "In practice these would be configurable" like through an environment variable, or something else? (just checking if I understand)

Author: Yea I think that configuration in this first iteration of plugins is up to the plugin's creator. So in most cases I would imagine an environment variable, but in the RFC one of the potentials is like a class that can be instantiated from your own plugin with props passed in. Maybe even a config file exists. I added some of those as possibilities for clarity maybe.

```js
// exports the postSynth function
exports.postSynth = function(config) {
  config.terraform.backend = {
    s3: {
      bucket: "my-wing-state-bucket",
      key: "plugins-rock.tfstate",
      region: "us-east-1",
    }  
  }
  return config;
}
```

### `validate` stage

> This stage is called right after the `postSynth` stage and provided the same context object. In the context of a cdktf target, this is the same terraform JSON configuration. However, this stage is immutable which allows for validation of the configuration without worrying about another plugin mutating after the fact.

Reviewer: ```suggestion
This stage is called right after the `postSynth` stage and provided the same context object. In the context of a terraform-based target like `tf-aws`, this is the same Terraform JSON configuration. However, this stage does not allow configuration to be mutated, which allows plugins to validate the configuration without worrying about another plugin mutating after the fact.
```

> This stage is called right after the `postSynth` stage and provided the same context object. In the context of a terraform-based target like `tf-aws`, this is the same Terraform JSON configuration. However, does not allow configuration to be mutated, which allows plugins to validate the configuration without worrying about another plugin mutating after the fact.

Reviewer 2: Please fix "terraform" to "Terraform" across the board.
